third_party/compilerworks/vertica_to_bq_migration/regexp_extract_with_occurrence_and_flags.sql

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2013-2020 CompilerWorks
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *     * Redistributions of source code must retain the above copyright notice,
+ *     this list of conditions and the following disclaimer.
+ *
+ *     * Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/* Internal function */
+CREATE OR REPLACE FUNCTION cw_regex_mode (mode STRING) RETURNS STRING
+LANGUAGE js AS """
+  var m = '';
+  if (mode == 'i' || mode == 'm')
+    m += mode;
+  else if (mode == 'n')
+    m += 's';
+  m += 'g';
+  return m;
+""";
+
+
+CREATE OR REPLACE FUNCTION cw_regexp_substr_generic (str STRING, regexp STRING, p INT64, o INT64, mode STRING) RETURNS STRING
+LANGUAGE js AS """
+  if (str == null || regexp == null || p == null || o == null || mode == null) return null;
+  var r = new RegExp(regexp, mode);
+  var m = str.substring(p - 1).match(r);
+  if (m == null) return null;
+  return m[o - 1];
+""";
+
+CREATE OR REPLACE FUNCTION regexp_extract_with_occurrence_and_flags (h STRING, n STRING, p INT64, o INT64, mode STRING) RETURNS STRING AS
+(
+    cw_regexp_substr_generic(h, n, p, o, cw_regex_mode(mode))
+);
```

Comment:
Split each into their own file. This will simplify unit testing.

Reply:
Done